### PR TITLE
[#1019] Async SDK

### PR DESCRIPTION
**Pull request by vlad-tokarev**
* Refactored _request API for EdiClient in SDK
* Added layer of async API for some sdk methods
* Integrate jupyter with async SDK

### CODE REVIEW

File: legion/sdk/legion/sdk/clients/edi.py [hunk 1]
```diff
@@ -95,6 +98,12 @@ def get_authorization_redirect(web_redirect: str, after_login: Callable) -> str:
     return web_redirect
 
 
+def get_prepared_request(*args, **kwargs):
```
Comment (mcdoker18) [inline legion/sdk/legion/sdk/clients/edi.py]: Add a type hinting for all your new functions

File: legion/sdk/legion/sdk/clients/edi_aggregated.py [hunk 2]
```diff
@@ -196,14 +227,16 @@ def parse_resources_file_with_one_item(path: str) -> LegionCloudResourceUpdatePa
     return resources.changes[0]
 
 
-def apply(updates: LegionCloudResourcesUpdateList, edi_client: RemoteEdiClient, is_removal: bool) -> ApplyResult:
+async def apply(updates: LegionCloudResourcesUpdateList,
```
Comment (mcdoker18) [inline legion/sdk/legion/sdk/clients/edi_aggregated.py]: Did you try to run the robot tests?
We use this function in the CLI package. It doesn't expect to get a coroutine.

File: legion/sdk/legion/sdk/clients/edi.py [hunk 3]
```diff
@@ -331,3 +321,210 @@ def info(self):
...
+                LOGGER.error('Failed to connect to {}: {}. Retrying'.format(self._base_url, exception))
+                raised_exception = exception
+            else:
+                LOGGER.debug('Got response. Breaking')
+                break
+            left_retries -= 1
+        else:
+            raise EDIConnectionException('Can not reach {}'.format(self._base_url)) from raised_exception
+
+        if self._login_required(response):
```
Comment (mcdoker18) [inline legion/sdk/legion/sdk/clients/edi.py:464]: All our security functions, which interacts with a Keycloak, are synchronous. Is it expected?
`legion/jyputerlab/handlers/oauth2.py` is too.

**Review by mcdoker18 — CHANGES_REQUESTED** [review]

File: legion/sdk/legion/sdk/clients/edi.py (hunk 3 above)
Comment (vlad-tokarev) [reply legion/sdk/legion/sdk/clients/edi.py:464]: Yes, I supposed to add async features iteratively. It allows not to rewrite all codebase at once. Therefore I optimized only SDK methods that affect jupyter performance

**Pull request closed without merging**